From b3d609b4da8f84356eb1c3933c291b13b77ebda5 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 28 Nov 2025 13:33:24 +0100
Subject: [PATCH] firmware: arm_scmi: support timeout without timekeeping

The ktime API used for timeout in shmem_tx_prepare() and
wait_for_completion_timeout API in scmi_wait_for_reply() requires
support of timekeeping but it is stopped during system suspend phases,
so the possible SCMI requests in drivers to stop clocks or regulators
sent when timekeeping_suspended = 1 cause an error in ktime_get()
or in schedule() [__schedule_bug].

To solve the issue, the spin_until_cond() with ktime functions
is replaced with call to readl_relaxed_poll_timeout_atomic(), because this
macro does not rely on timekeeping, and the polling mode is forced when
timekeeping is suspended.

This problem occurs on STM32MP2 platform with SCMI server in TF-M
based on IPCC mailbox when LTIMER driver stop/unprepare its clocks
with the call stack:

[   44.778882]  ktime_get+0x94/0x9c
[   44.778882]  shmem_tx_prepare+0x24/0x150
[   44.778882]  tx_prepare+0x14/0x20
[   44.778882]  msg_submit+0x8c/0x118
[   44.778882]  mbox_send_message+0xb4/0x130
[   44.778882]  mailbox_send_message+0x38/0x78
[   44.778882]  do_xfer+0xb4/0x21c
[   44.778882]  scmi_clock_config_set+0xa8/0xfc
[   44.778882]  scmi_clock_disable+0x18/0x24
[   44.778882]  scmi_clk_disable+0x24/0x30
[   44.778882]  clk_core_unprepare+0x8c/0x110
[   44.778882]  clk_core_unprepare+0x94/0x110
[   44.778882]  clk_unprepare+0x2c/0x44
[   44.778882]  stm32_clkevent_lp_suspend+0x2c/0x3c
[   44.778882]  clockevents_suspend+0x44/0x64
[   44.778882]  timekeeping_suspend+0x228/0x2f4
[   44.778882]  syscore_suspend+0x58/0x13c
[   44.778882]  suspend_devices_and_enter+0x42c/0x550
[   44.778882]  pm_suspend+0x1ac/0x260

Change-Id: Ic8fd169b9d4a9fbed5cd139db67b5d6993e374d9
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/505721
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/firmware/arm_scmi/common.h |  3 +-
 drivers/firmware/arm_scmi/driver.c | 45 ++++++++++++++++++++++--------
 drivers/firmware/arm_scmi/shmem.c  | 14 +++++-----
 3 files changed, 43 insertions(+), 19 deletions(-)

diff --git a/drivers/firmware/arm_scmi/common.h b/drivers/firmware/arm_scmi/common.h
index dc95652fff40..c24e24428996 100644
--- a/drivers/firmware/arm_scmi/common.h
+++ b/drivers/firmware/arm_scmi/common.h
@@ -261,7 +261,8 @@ struct scmi_desc {
 static inline bool is_polling_required(struct scmi_chan_info *cinfo,
 				       const struct scmi_desc *desc)
 {
-	return cinfo->no_completion_irq || desc->force_polling;
+	return cinfo->no_completion_irq || desc->force_polling ||
+	       timekeeping_suspended;
 }
 
 static inline bool is_transport_polling_capable(const struct scmi_desc *desc)
diff --git a/drivers/firmware/arm_scmi/driver.c b/drivers/firmware/arm_scmi/driver.c
index fbe893734411..2e7022f4a223 100644
--- a/drivers/firmware/arm_scmi/driver.c
+++ b/drivers/firmware/arm_scmi/driver.c
@@ -18,6 +18,7 @@
 
 #include <linux/bitmap.h>
 #include <linux/debugfs.h>
+#include <linux/delay.h>
 #include <linux/device.h>
 #include <linux/export.h>
 #include <linux/idr.h>
@@ -1058,8 +1059,7 @@ static void xfer_put(const struct scmi_protocol_handle *ph,
 }
 
 static bool scmi_xfer_done_no_timeout(struct scmi_chan_info *cinfo,
-				      struct scmi_xfer *xfer, ktime_t stop,
-				      bool *ooo)
+				      struct scmi_xfer *xfer, bool *ooo)
 {
 	struct scmi_info *info = handle_to_scmi_info(cinfo->handle);
 
@@ -1068,7 +1068,14 @@ static bool scmi_xfer_done_no_timeout(struct scmi_chan_info *cinfo,
 	 * in case of out-of-order receptions of delayed responses
 	 */
 	return info->desc->ops->poll_done(cinfo, xfer) ||
-	       (*ooo = try_wait_for_completion(&xfer->done)) ||
+	       (*ooo = try_wait_for_completion(&xfer->done));
+}
+
+static bool scmi_xfer_done_timeout(struct scmi_chan_info *cinfo,
+				   struct scmi_xfer *xfer, ktime_t stop,
+				   bool *ooo)
+{
+	return scmi_xfer_done_no_timeout(cinfo, xfer, ooo) ||
 	       ktime_after(ktime_get(), stop);
 }
 
@@ -1087,14 +1094,30 @@ static int scmi_wait_for_reply(struct device *dev, const struct scmi_desc *desc,
 		if (!desc->sync_cmds_completed_on_ret) {
 			bool ooo = false;
 
-			/*
-			 * Poll on xfer using transport provided .poll_done();
-			 * assumes no completion interrupt was available.
-			 */
-			ktime_t stop = ktime_add_ms(ktime_get(), timeout_ms);
-
-			spin_until_cond(scmi_xfer_done_no_timeout(cinfo, xfer,
-								  stop, &ooo));
+			if (timekeeping_suspended) {
+				/* Poll with when timekeeping not available */
+				unsigned int retries = timeout_ms * 100;
+
+				while (retries--) {
+					if (scmi_xfer_done_no_timeout(cinfo,
+								      xfer,
+								      &ooo))
+						break;
+					udelay(10);
+				}
+			} else {
+				/*
+				 * Poll on xfer using transport provided .poll_done();
+				 * assumes no completion interrupt was available.
+				 */
+				ktime_t stop = ktime_add_ms(ktime_get(),
+							    timeout_ms);
+
+				spin_until_cond(scmi_xfer_done_timeout(cinfo,
+								       xfer,
+								       stop,
+								       &ooo));
+			}
 			if (!ooo && !info->desc->ops->poll_done(cinfo, xfer)) {
 				dev_err(dev,
 					"timed out in resp(caller: %pS) - polling\n",
diff --git a/drivers/firmware/arm_scmi/shmem.c b/drivers/firmware/arm_scmi/shmem.c
index 517d52fb3bcb..e74c092644dc 100644
--- a/drivers/firmware/arm_scmi/shmem.c
+++ b/drivers/firmware/arm_scmi/shmem.c
@@ -7,6 +7,7 @@
 
 #include <linux/ktime.h>
 #include <linux/io.h>
+#include <linux/iopoll.h>
 #include <linux/processor.h>
 #include <linux/types.h>
 
@@ -35,7 +36,8 @@ struct scmi_shared_mem {
 void shmem_tx_prepare(struct scmi_shared_mem __iomem *shmem,
 		      struct scmi_xfer *xfer, struct scmi_chan_info *cinfo)
 {
-	ktime_t stop;
+	int ret;
+	u32 val;
 
 	/*
 	 * Ideally channel must be free by now unless OS timeout last
@@ -51,12 +53,10 @@ void shmem_tx_prepare(struct scmi_shared_mem __iomem *shmem,
 	 * this is just to ease debugging and avoid complete hangs on boot
 	 * due to a misbehaving SCMI firmware.
 	 */
-	stop = ktime_add_ms(ktime_get(), 2 * cinfo->rx_timeout_ms);
-	spin_until_cond((ioread32(&shmem->channel_status) &
-			 SCMI_SHMEM_CHAN_STAT_CHANNEL_FREE) ||
-			 ktime_after(ktime_get(), stop));
-	if (!(ioread32(&shmem->channel_status) &
-	      SCMI_SHMEM_CHAN_STAT_CHANNEL_FREE)) {
+	ret = readl_relaxed_poll_timeout_atomic(&shmem->channel_status, val,
+						val & SCMI_SHMEM_CHAN_STAT_CHANNEL_FREE,
+						0, 2000 * cinfo->rx_timeout_ms);
+	if (ret < 0) {
 		WARN_ON_ONCE(1);
 		dev_err(cinfo->dev,
 			"Timeout waiting for a free TX channel !\n");
-- 
2.53.0

